cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/deploy/pipeline/LocalArchiveLoaderStage.java

```diff
@@ -60,11 +68,33 @@ public LocalArchiveLoaderStage(CConfiguration cConf, Id.Namespace id, @Nullable
    * Creates a {@link co.cask.cdap.internal.app.deploy.InMemoryConfigurator} to run through
    * the process of generation of {@link ApplicationSpecification}
    *
-   * @param archive Location of archive.
+   * @param deploymentInfo Location of the input and output location
    */
   @Override
-  public void process(Location archive) throws Exception {
-    InMemoryConfigurator inMemoryConfigurator = new InMemoryConfigurator(id, archive);
+  public void process(DeploymentInfo deploymentInfo) throws Exception {
+
+    Location outputLocation = deploymentInfo.getDestination();
+    Location parent = Locations.getParent(outputLocation);
+    Locations.mkdirsIfNotExists(parent);
+
+    File input = deploymentInfo.getInputLocation();
+    Location tmpLocation = parent.getTempFile(".tmp");
+    LOG.debug("Copy from {} to {}", input.getName(), tmpLocation.toURI());
+    Files.copy(input, Locations.newOutputSupplier(tmpLocation));
+
+    // Finally, move archive to final location
+    try {
+      if (tmpLocation.renameTo(outputLocation) == null) {
+        throw new IOException(String.format("Could not move archive from location: %s, to location: %s",
+                                            tmpLocation.toURI(), outputLocation.toURI()));
+      }
+    } catch (IOException e) {
```

Comment:
Since we are doing a rename (move - which is atomic), we will need to do a delete only when rename fails. The only exception that is thrown on renameTo is IOException. So this would be good as is.